### PR TITLE
[D2] Assign and verify deterministic receipt sections

### PR DESCRIPTION
**Pull request by tnorlund**
## Summary

- Persist canonical `ReceiptRow` records on all three ingest paths, replacing only stale rows for the exact receipt while preserving LayoutLM outputs.
- Replace the fitted numeric-only section model with QA-measured row evidence: amount-column/context density, quantity/unit-price tokens, global lexical fallback, measured run transitions, and a semi-Markov decoder that permits repeated/non-monotone sections.
- Keep proposals additive and PENDING, protect human-VALID evidence from demotion/overlap, and share canonical Chroma/glyphstudio propagation primitives.
- Make the priors artifact byte-deterministic and auditable by omitting wall-clock metadata and hashing the complete labeled feature + normalized merchant corpus.
- Pin QA-backed ITEMS rows for the required Chick-fil-A and Smith's real-receipt goldens.

## Measured dev evidence

Read-only evaluation against the same 22 QA-valid held-out receipts:

- Overall agreement: **365/429 (0.8508)**; required >= 0.80
- ITEMS recall: **47/58 (0.8103)**; required >= 0.70
- Unassigned rows: **0**
- Chick-fil-A `cb581977`: 16/25, with four QA ITEMS rows pinned
- Smith's `ff574b98`: 22/23, with three QA ITEMS rows pinned

Training used 820 other receipts and excluded all 22 holdouts before learning. A fresh independent rebuild passed at the same 365/429 and 47/58; global-only inference also passed overall at 355/429, so merchant priors are not fitted to the acceptance cohort.

- Exclusion manifest SHA-256: `16fbda1597fe69060a65154b8ddc9205c4a4cd739085409dc31286431263c741`
- Section snapshot SHA-256: `9ce1e2308d07b77dbb64390cc202f2915d56d3ffa97e5d51ca08c31e94334415`
- Training corpus SHA-256: `4fdcfc97078cec2d996263e4bc7f238e4e37d1f41a1145cb652918884ee16cd4`
- Packaged priors SHA-256: `9e6e828a83b889f20d105ab2fa2864fc217bd7c80a00e0176cfaec36764fb2b2`

Both builder and evaluator refuse every table except exact dev `ReceiptsTable-dc5be22` and perform no DynamoDB writes. No `.npz` is committed; no LayoutLM label deletion is introduced.

## Verification

- `receipt_upload` full suite: pass (existing documented skips only)
- `receipt_chroma`: 492 passed, 7 skipped
- `receipt_dynamo`: 4,271 passed; one unrelated pre-existing `ReceiptMetadata` address-quality test fails identically on `origin/main` (this branch changes only `ReceiptRow`/`ReceiptSection` there)
- OCR handler overlay: 54 passed
- CodeBuild context/hash contracts: 8 passed
- Exact CI-pinned Black 26.5.1 and isort 8.0.1: pass for changed packages
- glyphstudio propagation: 6 passed; built wheel declares `receipt-chroma`
- Receipt-upload wheel contains the exact committed priors artifact
- Independent read-only audits: no remaining P0/P1 findings

The earlier ARM64 image build/import smoke passed. The single allowed final rebuild attempt did not execute the Dockerfile because public ECR returned 403 while resolving `public.ecr.aws/lambda/python:3.12`; no repository workaround was added.

## Prerequisite state

The documented human chain remains external: #1154 and #1146 are clean/green but open; #1152 is open with the inherited `receipt_dynamo` failure. This PR is based directly on `main` and remains draft pending normal CI/review.

### CODE REVIEW

File: receipt_upload/receipt_upload/section_verifier.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        else:
+            predicted = None
+            status = "ABSTAINED"
+        updated = replace(
+            section,
+            validation_status=(
+                ValidationStatus.PENDING.value
+                if disagreements
+                else section.validation_status
+            ),
```
Comment (tnorlund) [inline receipt_upload/receipt_upload/section_verifier.py:217]: **P2 — verifier can demote a human-QA'd VALID section back to PENDING.** `_record_verification` selects sections purely by `model_source == MODEL_SOURCE` (line 175) and, on disagreement, unconditionally sets `validation_status = PENDING` regardless of the *current* status. A section created here retains `model_source=upload-determinism-v1` even after the QA workflow promotes it to VALID. So a later re-verify (re-embed / drain re-run — which this project does at scale) whose KNN disagrees will silently regress a human VALID → PENDING. The corpus is 91% VALID ground truth; the spec's "never overwrites" / "disagreement -> PENDING never silent override" is about machine-PENDING rows, not human verdicts. Guard it: only demote when `section.validation_status == ValidationStatus.PENDING.value` (still record the DISAGREED provenance on VALID rows, just don't flip the status).

File: receipt_upload/receipt_upload/section_verifier.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            verification_section_type=predicted,
+            verification_confidence=(
+                fmean(item.confidence for item in available)
+                if available
+                else None
+            ),
+            disagreement_row_ids=disagreements or None,
+            verified_at=now,
+        )
+        dynamo.update_receipt_section(updated)
```
Comment (tnorlund) [inline receipt_upload/receipt_upload/section_verifier.py:229]: **P3 — verifier rewrites every upload-determinism section on every run, even when nothing changed.** The AGREED/ABSTAINED branches still `replace(..., verified_at=now, ...)` and call `update_receipt_section`, so `verified_at` always differs and every section is a fresh write → a RECEIPT_SECTION stream event → Chroma metadata sync, on each re-embed. Bounded per receipt (~8 sections) so not a storm, but during a full corpus re-embed it's pure churn. Consider skipping the write when the verification outcome + status are unchanged (compare against the read section), or only bumping `verified_at` when a field actually changes.

File: receipt_upload/receipt_upload/section_verifier.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+    if not rows or not row_embeddings:
+        return []
+    if len(rows) != len(row_embeddings):
+        raise ValueError("rows and row_embeddings must have equal length")
+    result = chroma.query(
+        collection_name="lines",
+        query_embeddings=[list(embedding) for embedding in row_embeddings],
+        n_results=KNN_NEIGHBORS,
+        include=["embeddings", "metadatas"],
```
Comment (tnorlund) [inline receipt_upload/receipt_upload/section_verifier.py:126]: **P3 — verification recall can starve on re-embeds.** `n_results=KNN_NEIGHBORS` (15) is fetched, then same-receipt neighbors are filtered out at lines 130-134. On a re-embed the receipt's own rows are already in the `lines` snapshot and are the *nearest* neighbors, so the top-15 can be dominated by same-receipt rows → few or zero cross-receipt training examples → `continue` (abstain) for most rows. Either request more neighbors (e.g. `n_results` well above 15) or push the exclusion into the query via a Chroma `where` clause on `image_id`, so the 15 you keep are already cross-receipt.

File: receipt_upload/receipt_upload/section_assignment.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    tokens = section_model.get("tokens", {})
+    unknown = float(section_model.get("unknown_token_probability", _EPSILON))
+    if features.tokens and "tokens" in section_model:
+        scores.append(
+            fmean(
+                math.log(float(tokens.get(token, unknown)))
+                for token in features.tokens
+            )
+        )
+    return fmean(scores)
```
Comment (tnorlund) [inline receipt_upload/receipt_upload/section_assignment.py]: **P3 — repeat-merchant priors silently drop all token/keyword evidence.** The build script fits merchant priors with `include_tokens=False`, so merchant `section_models` have no `tokens` key. Here the token branch is gated on `"tokens" in section_model` and reads only `section_model`'s tokens (never the global `fallback_model`'s). Net effect: a receipt from one of the 93 repeat merchants is scored on the 6 numeric Gaussians alone, while an unknown-merchant receipt (global prior) also gets keyword evidence — the higher-signal merchants get the *weaker* emission model. If that asymmetry is intentional (avoid merchant token overfit), a one-line comment would save the next reader; otherwise fall back to `fallback_model`'s tokens when the merchant model lacks them.

File: receipt_upload/tests/test_section_assignment.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    for fixture in fixtures:
+        rows, lines = _rows(fixture["rows"])
+        first = assign_row_sections(rows, lines, model, fixture["merchant"])
+        second = assign_row_sections(rows, lines, model, fixture["merchant"])
+        assert first == second
+        assert len(first) == len(rows)
+        assert all(
+            item.section_type in SectionType._value2member_map_
+            for item in first
+        )
```
Comment (tnorlund) [inline receipt_upload/tests/test_section_assignment.py:293]: **P2 — the "golden" fixtures don't assert any expected section, only determinism + valid-enum.** `test_golden_merchants_have_deterministic_learned_priors` checks `first == second` and that each type is a real `SectionType`. That can't catch an assignment-quality regression: Vons `MEMBER SAVINGS 1.00`, Smith's `FUEL POINTS EARNED 3`, or Italia Deli's ordering (the explicitly-requested weird-order case) could all be mis-sectioned and the test still passes. Add per-row expected `section_type` assertions to the fixture (at least the anchors: storefront row → STOREFRONT, total row → TOTAL_LINE/SUMMARY, item row → ITEMS) so these goldens actually pin behavior.

File: scripts/build_section_order_priors.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    if args.environment != "dev":
+        raise SystemExit(
+            "Refusing to read anything except the dev environment"
+        )
+    if not table_name:
+        raise SystemExit("DYNAMODB_TABLE_NAME is required")
+    client = DynamoClient(table_name)
+    sections, cursor = client.list_receipt_sections()
+    if cursor is not None:
+        raise RuntimeError("Unexpected unconsumed section pagination cursor")
```
Comment (tnorlund) [inline scripts/build_section_order_priors.py]: **P3 — single-page assumption on `list_receipt_sections()`.** `sections, cursor = client.list_receipt_sections(); if cursor is not None: raise`. This fails loud (good) but aborts the whole build the moment the section corpus outgrows one Scan page. It worked at 5,593 sections, but that's near the 1MB page boundary for these rows. Loop the cursor and accumulate rather than refusing, so the builder keeps working as the QA corpus grows.

File: receipt_upload/receipt_upload/merchant_resolution/embedding_processor.py [hunk 7]
```diff
@@ -497,11 +500,63 @@ def _do_lines_work() -> Dict[str, Any]:
...
+            )
+
+            rows_by_id = {row.row_id: row for row in persisted_rows}
+            embedding_rows = [
+                rows_by_id[record.primary_line.line_id]
+                for record in row_records
+            ]
+            verification_stats: Dict[str, Any] = {}
+            try:
+                verified = verify_receipt_sections(
```
Comment (tnorlund) [inline receipt_upload/receipt_upload/merchant_resolution/embedding_processor.py:526]: **P3 — the "async" verifier actually runs synchronously on the upload critical path.** The spec described the KNN verifier as an async pass that upgrades when embeddings land; here it's invoked inline in the lines worker (an extra `chroma.query` + per-row KNN) right after assignment. Functionally fine and it's correctly wrapped so a failure can't discard the deterministic proposal — but it adds latency to every upload and couples verification to the snapshot being present. Worth a note in the docstring that this is the sync path, and consider whether it belongs behind the same kind of feature gate as the LLM async hand-off if upload latency is sensitive.

**Review by tnorlund — COMMENTED** [review]
## D2 review — deterministic section assignment + KNN verify

**Verdict: solid, mergeable after addressing one QA-safety guard. No P1 blockers.** Ran the focused suite (test_section_assignment, test_section_verifier, test_receipt_section, receipt_chroma public API) against the worktree — 68 passing.

Severity counts: **P1: 0 · P2: 2 · P3: 5**

### What the review confirmed is correct
- **Operates on the merged spine, not raw lines.** `assign_row_sections` consumes `ReceiptRow`s, features come from row geometry, and sections carry `row_ids` (row-granular). The embedding_processor reads persisted D1 rows (`get_receipt_rows_from_receipt`) and only falls back to `build_receipt_rows` for legacy/dev replays. Good.
- **Creation is additive; VALID/QA'd sections are never overwritten at assign time.** `assign_and_persist_sections` skips any `section_type` that already exists (any status) and catches `EntityAlreadyExistsError`. Assignment runs before the first lines delta so `section_by_line` reaches `build_row_payload`. Good.
- **SK-per-type uniqueness respected.** The monotone Viterbi guarantees each `section_type` is contiguous, `sections_from_assignments` aggregates one section per type, and the DynamoDB key is `SECTION#{type}`. One row per (receipt, type). Good.
- **Prior learning is VALID-only, per-merchant, no hardcoding.** `build_model` filters `ValidationStatus.VALID`, the repeat-merchant cutoff is derived (`> corpus median`, 93 merchants), and the snapshot SHA-256 + counts are recorded rather than asserted. Good.
- **Verifier annotates VALID neighbors only and never changes `section_type` or row membership.** `_valid_line_sections` filters VALID; disagreement records provenance + demotes to PENDING. Matches the spec intent for machine-PENDING rows.

The two P2s and the P3s are inline. Nothing here is a data-loss risk; the QA-demotion guard (P2) is the one I'd want addressed before this runs against the 91%-VALID corpus.
